builtins/web/fetch/headers.cpp

```diff
@@ -923,12 +1079,188 @@ unique_ptr<host_api::HttpHeaders> Headers::handle_clone(JSContext* cx, HandleObj
+         JS_DeleteProperty(cx, proto_obj, "constructor");
+}
+
+std::tuple<host_api::HostString, host_api::HostString> *
+Headers::get_index(JSContext *cx, JS::HandleObject self, size_t index) {
+  MOZ_ASSERT(is_instance(self));
+  std::vector<size_t> *headers_sort_list = static_cast<std::vector<size_t> *>(
+      JS::GetReservedSlot(self, static_cast<size_t>(Headers::Slots::HeadersSortList)).toPrivate());
+  MOZ_ASSERT(headers_sort_list);
+
+  HeadersList *headers_list = static_cast<Headers::HeadersList *>(
+      JS::GetReservedSlot(self, static_cast<size_t>(Headers::Slots::HeadersList)).toPrivate());
+  MOZ_ASSERT(headers_list);
+
+  ensure_updated_sort_list(headers_list, headers_sort_list);
+  MOZ_RELEASE_ASSERT(index < headers_sort_list->size());
+
+  return &headers_list->at(headers_sort_list->at(index));
+}
+
+std::optional<size_t> Headers::lookup(JSContext *cx, HandleObject self, string_view key) {
+  MOZ_ASSERT(is_instance(self));
+  const HeadersList *headers_list = static_cast<Headers::HeadersList *>(
+      JS::GetReservedSlot(self, static_cast<size_t>(Headers::Slots::HeadersList)).toPrivate());
+  MOZ_ASSERT(headers_list);
+  std::vector<size_t> *headers_sort_list = static_cast<std::vector<size_t> *>(
+      JS::GetReservedSlot(self, static_cast<size_t>(Headers::Slots::HeadersSortList)).toPrivate());
+  MOZ_ASSERT(headers_sort_list);
+
+  ensure_updated_sort_list(headers_list, headers_sort_list);
+
+  // Now we know its sorted, we can binary search.
+  auto it = std::lower_bound(headers_sort_list->begin(), headers_sort_list->end(), key,
+                             HeadersSortListLookupCompare(headers_list));
+  if (it == headers_sort_list->end() ||
+      header_compare(std::get<0>(headers_list->at(*it)), key) != Ordering::Equal) {
+    return std::nullopt;
+  }
+  return it - headers_sort_list->begin();
+}
+
+bool HeadersIterator::next(JSContext *cx, unsigned argc, Value *vp) {
+  METHOD_HEADER(0)
+  JS::RootedObject headers(cx, &JS::GetReservedSlot(self, Slots::Headers).toObject());
+
+  Headers::HeadersList *list = Headers::get_list(cx, headers);
+
+  size_t index = JS::GetReservedSlot(self, Slots::Cursor).toInt32();
+  size_t len = list->size();
+  uint8_t type = static_cast<uint8_t>(JS::GetReservedSlot(self, Slots::Type).toInt32());
+
+  JS::RootedObject result(cx, JS_NewPlainObject(cx));
+  if (!result)
+    return false;
+
+  if (index >= len) {
+    JS_DefineProperty(cx, result, "done", JS::TrueHandleValue, JSPROP_ENUMERATE);
+    JS_DefineProperty(cx, result, "value", JS::UndefinedHandleValue, JSPROP_ENUMERATE);
+
+    args.rval().setObject(*result);
+    return true;
+  }
+
+  JS_DefineProperty(cx, result, "done", JS::FalseHandleValue, JSPROP_ENUMERATE);
+
+  JS::RootedValue key_val(cx);
+  JS::RootedValue val_val(cx);
+
+  // If there was a header prepend, we must recompute the distinct index for proper iteration.
+  // We detect prepends by checking the last_key sorting.
```

Comment:
Writing this comment for myself - this comment is out of date from a former version of this code.